### DISCUSSION TITLE
Adding rule definition to helm chart

### DISCUSSION THREAD

**Discussion by nunogoncalocastro-vodafone** [emqx/emqx#13968] [Q&A] [answered]
We are trying to deploy an EMQX broker in a Kubernetes cluster, and including all the required configurations on the EMQX.yaml file.
Currently the only missing part (which is still being included manually) is the Rule Definition for a Data Bridge feature.

We retrieved the rule definition from the inside the running pod and it looked like this
```
       rule_engine {
         ignore_sys_message = true
         jq_function_default_timeout = "10s"
         rules {
           rule_bridge {
             actions = [
               "mqtt:hivemq"
             ]
             description = ""
             metadata {created_at = 1726764090774}
             sql = "~
               SELECT 
                 payload as payload_original,
                 hexstr2bin( substr( bin2hexstr( payload), 136 )) as payload, 
                 topic, clientid,
                 ('prefix/' + replace(topic,'mqttsn/','/') + clientid) as new_topic
               FROM
                 "mqttsn/#"~"
           }
         }
       }
```

When we tried to incorporate that into the Config.Data section of our EMQX.yaml file, it was failing and pointing to the use of "~" in the Rule SQL attribute.

We searched for similar cases and none seemed to contain that ~ character, and were replacing the apostrophe (') with the escaped quotation mark (\") instead, which we also did.

At this point the pod didn't start and the emqx-operator logs was referring to a forbidden character in line 139
      \"(\" is a forbidden character in keys\nfailed to parse config
which refers to the line
      hexstr2bin( substr( bin2hexstr( payload), 136 )) as payload, 

If we can't use "(", how can we apply the functions on the SELECT statement? or it requires escape character before?

**Comment by thalesmg** [comment]
Hi.

> When we tried to incorporate that into the Config.Data section of our EMQX.yaml file, it was failing and pointing to the use of "~" in the Rule SQL attribute.

In the provided snippet, I see:

```
             sql = "~
...
               FROM
                 "mqttsn/#"~"
```

Which should be

```
             sql = "
...
               FROM
                 "mqttsn/#"
```

**Reply by nunogoncalocastro-vodafone** [reply]
Hi @thalesmg 
We tried different variations, but the current issue seems to be related with the use of parenthesis ( ) in the expression. 
```
           hexstr2bin( substr( bin2hexstr( payload), 136 )) as payload, 
```
At least that's our understanding of the received error message in the EMQX Operator logs.

This current error is showing after we removed the "~" and replaced the apostrophes ' with escaped quotations marks \\"
```
    sql = "
    SELECT
    ...
    FROM
        \"mqttsn/#\"  "
```
We also tried compacting everything in a single line, similar to other examples we saw in other posts, but the result is the same.
```
    sql = "SELECT\n ...\n  FROM\n \"mqttsn/#\"  "
```

**Reply by thalesmg** [reply]
Can you try as @id said below and put everything in a single line, without the `\n` newlines inside the string?

**Comment by id** [comment, marked as answer]
try this
```
rule_engine {
         ignore_sys_message = true
         jq_function_default_timeout = "10s"
         rules {
           rule_bridge {
             actions = [
               "mqtt:hivemq"
             ]
             description = ""
             metadata {created_at = 1726764090774}
             sql = "SELECT payload as payload_original, hexstr2bin( substr( bin2hexstr( payload), 136 )) as payload, topic, clientid, ('prefix/' + replace(topic,'mqttsn/','/') + clientid) as new_topic FROM 'mqttsn/#'"
           }
         }
       }
```

**Reply by nunogoncalocastro-vodafone** [reply]
Thanks @id and @thalesmg , that did the trick.
So no double quotation marks, escape or tilde characters, just apostrophes inside the SQL expression.

**Comment by zmstone** [comment]
Hey.
Tilde `~` immediately following `“””` is for multi line strings with indentation.

> Multiline strings allow indentation (spaces, not tabs). If ~\n (or ~\r\n) are the only characters following the opening triple-quote, then it's a multiline string with indentation:
- The first line ~\n is discarded.
- The closing triple-quote can be either `"""` or `~"""` (`~` allows the string to end with " without escaping).
- Indentation is allowed but not required for empty lines.
- Indentation level is determined by the least number of leading spaces among the non-empty lines.

Ref here: https://github.com/emqx/hocon?tab=readme-ov-file#divergence-from-spec-and-caveats
